### PR TITLE
[12.0][IMP] project_role: hierarchy, archived roles and assignments

### PR DESCRIPTION
**Pull request by alexey-pelykh**
After few discussions, concept of roles has been improved to more cover needs of complex project task management.

### CODE REVIEW

**Comment by alexey-pelykh** [comment]
@mreficent @astirpe @pedrobaeza I believe you may be interested in this improvement

File: project_role/models/project_assignment.py [hunk 1]
```diff
@@ -18,50 +21,111 @@ class ProjectAssignment(models.Model):
...
-    @api.constrains('role_id', 'user_id')
-    def _check_assignable(self):
+    @api.constrains('company_id', 'project_id', 'role_id', 'user_id')
+    def _check(self):
         for assignment in self:
+            conflicting_assignment = self.search(
+                assignment._get_conflicting_domain(),
+                limit=1,
+            )
+            if conflicting_assignment:
```
Comment (tarteo) [inline project_role/models/project_assignment.py:122]: Isn't this redundant considering the constraint. Or do I misunderstand it :smile: maybe

**Review by tarteo — APPROVED** [review]
Nice addition / improvement! LGTM.

File: project_role/models/project_assignment.py (hunk 1 above)
Comment (alexey-pelykh) [reply project_role/models/project_assignment.py:122]: It's only a bit redundant, in SQL constraint it's hard (if even possible) to express logic like "if there's global assignment of user U to role R, don't allow project-specific assignment". So I could've removed SQL constraints but I think it's good to keep things consistent as much as possible on database level as well :)

**Review by rafaelbn — APPROVED** [review]

**Pull request merged**